calyx/src/passes/static_par_conv.rs

```diff
@@ -0,0 +1,141 @@
+//  is a Seq. Panics if Control statement is not a Seq.
+fn len_if_seq(stmt: &ir::Control) -> usize {
+    match stmt {
+        ir::Control::Seq(seq) => seq.stmts.len(),
+        _ => panic!("Not a sequence"),
+    }
+}
+
+// Given a Control statement and an index n.
+// If the Control statement is not a Seq, then panics.
+// Othrwise, returns Some(cycles), where cycles is the value of the "static"
+// attribute of the nth stmt in the Seq. Returns None if no such value is available.
+fn cycles_if_seq(s: &ir::Control, index: usize) -> Option<&u64> {
+    match s {
+        ir::Control::Seq(seq) => seq.stmts.get(index).and_then(|stmt| {
+            stmt.get_attributes().and_then(|atts| atts.get("static"))
+        }),
+        _ => panic!("Not a sequence"),
+    }
+}
+
+impl Visitor for StaticParConv {
+    fn finish_par(
+        &mut self,
+        s: &mut ir::Par,
+        _comp: &mut ir::Component,
+        _sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        //if the par block is empty we can stop
+        if (*s).stmts.is_empty() {
+            return Ok(Action::Continue);
+        }
+
+        //make sure only have seq stmts
+        if !(*s)
+            .stmts
+            .iter()
+            .all(|x| matches!(x, ir::Control::Seq(_seq)))
+        {
+            return Ok(Action::Continue);
+        }
+
+        //make sure each seq has the same number of stmts
+        let lens_vec =
+            (*s).stmts.iter().map(len_if_seq).collect::<Vec<usize>>();
+        let min_seq_len = *lens_vec
+            .iter()
+            .min()
+            .unwrap_or_else(|| panic!("empty par block"));
+        let max_seq_len = *lens_vec
+            .iter()
+            .max()
+            .unwrap_or_else(|| panic!("empty par block"));
+        if min_seq_len != max_seq_len {
+            return Ok(Action::Continue);
+        }
+        let seq_len = min_seq_len;
+
+        //make sure nth statement in each seq takes same number of cycle
+        for n in 0..seq_len {
+            let cycles_vec = (*s)
+                .stmts
+                .iter()
+                .map(|stmt| cycles_if_seq(stmt, n))
+                .collect::<Vec<Option<&u64>>>();
+            if cycles_vec.is_empty() {
+                panic!("empty par block");
+            }
+            let fst = cycles_vec[0];
+            if fst == None {
+                return Ok(Action::Continue);
+            }
+            if !cycles_vec.into_iter().all(|xth| xth == fst) {
+                return Ok(Action::Continue);
+            }
+        }
+
+        //It complains when I try to use the vec![] syntax. It says:
+        //the trait `Clone` is not implemented for `ir::control::Control`
+        //note: required by a bound in `from_elem`
+        //So I am doing it this way instead.
+        let mut new_seq_stmts = Vec::new();
+        for _n in 0..seq_len {
+            new_seq_stmts.push(Vec::new());
+        }
```

Comment:
Does `vec![vec![]; seq_len]` not work?

Reply:
here's the full error message when I try that: 

error[E0277]: the trait bound `ir::control::Control: Clone` is not satisfied
    --> calyx/src/passes/static_par_conv.rs:144:33
     |
144  |         let mut new_seq_stmts = vec![vec![]; max_seq_len];
     |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `ir::control::Control`
     |
     = note: required because of the requirements on the impl of `Clone` for `Vec<ir::control::Control>`
note: required by a bound in `from_elem`
    --> /Users/calebkim/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:2362:21
     |
2362 | pub fn from_elem<T: Clone>(elem: T, n: usize) -> Vec<T> {
     |                     ^^^^^ required by this bound in `from_elem`
     = note: this error originates in the macro `vec` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `calyx` due to previous error

Reply:
I stumbled across this relevant Rust forum thread:
https://users.rust-lang.org/t/how-to-construct-a-new-empty-vec-vec-t-with-specified-capacities/22937/2

In short, there may be some shorter ways to do it, but it appears there's no obviously better way to do it.